src/Microsoft.DotNet.Watcher.Tools/Internal/ProcessRunner.cs

```diff
@@ -0,0 +1,109 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Diagnostics;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Internal;
+using Microsoft.DotNet.Cli.Utils;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.DotNet.Watcher.Internal
+{
+    public class ProcessRunner
+    {
+        private readonly ILogger _logger;
+
+        public ProcessRunner(ILogger logger)
+        {
+            Ensure.NotNull(logger, nameof(logger));
+
+            _logger = logger;
+        }
+
+        public async Task<int> RunAsync(ProcessSpec processSpec, CancellationToken cancellationToken)
+        {
+            Ensure.NotNull(processSpec, nameof(processSpec));
+
+            var killLock = new object();
+            var tcs = new TaskCompletionSource<object>();
+            EventHandler handler = (sender, args) =>
+            {
+                tcs.TrySetResult(null);
+            };
+
+            var process = CreateProcess(processSpec);
+
+            process.Exited += handler;
+
+            var wasKilled = false;
+            var hasExited = false; // used to avoid race condition with process.Dispose()
+
+            cancellationToken.Register(() =>
+            {
+                if (!hasExited && !process.HasExited) 
+                {
+                    wasKilled = true;
+                    process.KillTree();
+                }
+                tcs.TrySetResult(null);
+            });
+
+            process.Start();
```

Comment:
Yeah, it can throw, but I don't think most of the conditions apply. https://msdn.microsoft.com/en-us/library/0w4h05yb(v=vs.110).aspx#Anchor_1. But if we start supporting arbitrary executable.